### DISCUSSION TITLE
Possible to revalidate sitemap.xml at runtime?

### DISCUSSION THREAD

**Discussion by fabiancdng** [vercel/next.js#50419] [App Router] [answered]
Hey everyone 👋

I love the new metadata capabilities of Next.js using the app router!

However, I've been wondering whether the sitemap (using the `sitemap()` generation function) supports some form of ISR or SSR...

https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generate-a-sitemap

As pages might be added at runtime in my application, that would be rather important...

Does anyone have any ideas/experience regarding this?

**Comment by viniciusbitt** [comment]
Running into this same issue now. But you can get it another way using [Route Handlers](https://nextjs.org/docs/app/building-your-application/routing/router-handlers#non-ui-responses):

app/sitemap.xml/route.ts
```ts
export async function GET(): Promise<Response> {
  const headers = new Headers();
  headers.set("Content-Type", "application/xml");

  return new Response(
    `<?xml version="1.0" encoding="UTF-8" ?>
        <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
            <url>
              <loc>https://acme.com</loc>
              <lastmod>${new Date().toISOString()}</lastmod>
              </url>
        </urlset>
    `,
    {
      headers,
    }
  );
}

export const revalidate = 5
```


I believed that because it was a normal route, the `sitemap generation function` could be `revalidate, dynamic and static`. I hope Next.js implements this.

**Reply by Galanthus** [reply]
@viniciusbitt Was banging my head against the wall... this helped me a lot thanks.

**Reply by FerrariAndrea** [reply]
In the old version of next.js (pages), we were able to set the revalidate time of the sitemap, I really don't understand why in the app version this is not possible anymore.
I will try to implement the 'app/sitemap.xml/route.ts' as @viniciusbitt is suggesting, thx man.

**Reply by lcnogueira** [reply]
There's a difference though in how those 2 approaches work: by using route handlers your page will be server side rendered. By using the nextjs `sitemap` file convention the page will be statically generated. I know this because I want to statically generate that sitemap but I'd need to revalidate it

**Reply by mdrayer** [reply]
Dang, almost a year and still no answers on if this will get implemented at some point or not.

**Reply by FerrariAndrea** [reply]
I have some issues opened on next.js, and yes usually they take a lot of time to answer because if you notice there are really a lot of issues, and after the release of version 14 I think they are really busy.
Anyway, the other way "app/sitemap.xml/route.js" with a revalidate seems to work fine as a temporary fix.

**Reply by lcnogueira** [reply]
I had answered this a few months ago (see comment above) and I said that by using route handlers the page would be server side rendered rather than statically generated. Turns out I was wrong...the reason why the sitemap files were being server side rendered (with streaming) was because I was using the edge runtime and it was disabling the static generation for the pages. I'm statically generating those pages with the _route.js_ now.

**Comment by mliudev** [comment]
Tried using `revalidatePaths('/sitemap.xml')` in a route handler to regenerate the static file generated by `sitemap.ts`, but that doesn't seem to work either.

**Comment by pawelhoros** [comment]
Next.js 14.2.3, just put in sitemap.ts:

`export const revalidate = 3600; // revalidate at most every hour`

and sitemap.xml is revalidated in production

**Reply by MartineBoulanger** [reply]
Can confirm this is working 😄

**Comment by fabiancdng** [comment, marked as answer]
Thanks to @pawelhoros's comment, I checked out the behavior of dynamic sitemaps in the newest Next.js version (14.2.3) and it seems like it is indeed possible to specify a revalidation interval in your `sitemap.ts`/`sitemap.js` file now.
You can check out my demo setup here: https://gist.github.com/fabiancdng/ef318c99d8a9f5efc2c48f6a06108881

If someone is still running into stale sitemaps despite the exported `revalidate` constant, I recommend upgrading to a newer Next.js version. If you don't want to (or can't) do that, I recommend the [solution proposed by](https://github.com/vercel/next.js/discussions/50419#discussioncomment-6189168) @viniciusbitt using route handlers.

Closing this discussion now.
Thanks to everyone who gave their input on this!

**Comment by Arctomachine** [comment]
Timed revalidation is not appropriate solution for sitemaps due to how it works. Because it wait for next visit to rebuild page. What this means in practice step by step:
1. You make changes to your site
2. New sitemap is needed, but not created yet
3. Crawler bot visits your site and checks sitemap
4. It finds that nothing changed and goes away without indexing
5. New sitemap is generated

The only realistic solution seems to be dynamic generation until manual revalidation is fixed

**Reply by employee451** [reply]
> The only realistic solution seems to be dynamic generation until manual revalidation is fixed

Or you can set up a cron job to hit sitemap.xml (https://cron-job.org/ for example)
